featuretools/computational_backends/feature_tree.py

```diff
@@ -153,9 +198,53 @@ def _get_feature_depths(self, entity_id):
 
         return list(features.values()), out
 
+    def _build_dependents(self):
+        feature_dependents = defaultdict(set)
+        for f in self.all_features:
+            dependencies = self.feature_dependencies[f.hash()]
+            for d in dependencies:
+                feature_dependents[d.hash()].add(f)
+        return feature_dependents
+
+    def _needs_all_values(self, feature):
+        if feature.needs_all_values:
+            return True
+        return self._dependent_needs_all_values(feature)
+
+    def _dependent_needs_all_values(self, feature):
+        for d in self.feature_dependents[feature.hash()]:
+            if d.needs_all_values:
+                return True
+        return False
 
 # These functions are used for sorting and grouping features
 
+    def needs_all_values_differentiator(self, f):
+        is_output = f.hash() in self.feature_hashes
+        # If a dependent feature requires all the instance values
+        # of the associated entity, then we need to calculate this
+        # feature on all values
+        # If the feature is one in which the user requested as
+        # an output (meaning it's part of the input feature list
+        # to calculate_feature_matrix), then we also need
+        # to subselect the output based on the desired instance ids
+        # and place in the return data frame.
+        if self._dependent_needs_all_values(f) and is_output:
```

Comment:
i think you can refactor this to get rid of extra functions

```
dependent_needs_all_values = False
for d in self.feature_dependents[f.hash()]:
        if d.needs_all_values:
            dependent_needs_all_values = True
needs_all_values = f.needs_all_values or dependent_needs_all_values
```
then use variables below